From f2434a25106df823e28f38e2d72ae5a972842c89 Mon Sep 17 00:00:00 2001
From: Anton Bambura <jenneron@protonmail.com>
Date: Thu, 16 Jun 2022 01:41:06 +0700
Subject: [PATCH] arm64: dts: qcom: add support for AYN Odin

Add device-trees for SDM845-based AYN Odin M0 and M2.

Based on initial implementation by Teguh Sobirin, extensively
reworked and extended since 2022.

TODO: Rename sdm845-ayn-odin to sdm845-ayn-odin-m0, sync with
pmaports changes

TODO: Ask Teguh Sobirin whether it is fine to arrange authorships
this way.

Co-developed-by: Teguh Sobirin <teguh@sobir.in>
Signed-off-by: Teguh Sobirin <teguh@sobir.in>
Signed-off-by: Anton Bambura <jenneron@protonmail.com>
---
 arch/arm64/boot/dts/qcom/Makefile             |    2 +
 .../boot/dts/qcom/sdm845-ayn-odin-common.dtsi | 1485 +++++++++++++++++
 .../boot/dts/qcom/sdm845-ayn-odin-m0.dts      |   57 +
 .../boot/dts/qcom/sdm845-ayn-odin-m2.dts      |   57 +
 arch/arm64/boot/dts/qcom/sdm845.dtsi          |   26 +-
 5 files changed, 1614 insertions(+), 13 deletions(-)
 create mode 100644 arch/arm64/boot/dts/qcom/sdm845-ayn-odin-common.dtsi
 create mode 100644 arch/arm64/boot/dts/qcom/sdm845-ayn-odin-m0.dts
 create mode 100644 arch/arm64/boot/dts/qcom/sdm845-ayn-odin-m2.dts

diff --git a/arch/arm64/boot/dts/qcom/Makefile b/arch/arm64/boot/dts/qcom/Makefile
index 557be6fd56d7..4561f1c3c806 100644
--- a/arch/arm64/boot/dts/qcom/Makefile
+++ b/arch/arm64/boot/dts/qcom/Makefile
@@ -303,6 +303,8 @@ dtb-$(CONFIG_ARCH_QCOM)	+= sdm632-motorola-ocean.dtb
 dtb-$(CONFIG_ARCH_QCOM)	+= sdm636-sony-xperia-ganges-mermaid.dtb
 dtb-$(CONFIG_ARCH_QCOM)	+= sdm660-xiaomi-lavender.dtb
 dtb-$(CONFIG_ARCH_QCOM)	+= sdm670-google-sargo.dtb
+dtb-$(CONFIG_ARCH_QCOM)	+= sdm845-ayn-odin-m0.dtb
+dtb-$(CONFIG_ARCH_QCOM)	+= sdm845-ayn-odin-m2.dtb
 dtb-$(CONFIG_ARCH_QCOM)	+= sdm845-db845c.dtb
 
 sdm845-db845c-navigation-mezzanine-dtbs	:= sdm845-db845c.dtb sdm845-db845c-navigation-mezzanine.dtbo
diff --git a/arch/arm64/boot/dts/qcom/sdm845-ayn-odin-common.dtsi b/arch/arm64/boot/dts/qcom/sdm845-ayn-odin-common.dtsi
new file mode 100644
index 000000000000..cd67aeaabae0
--- /dev/null
+++ b/arch/arm64/boot/dts/qcom/sdm845-ayn-odin-common.dtsi
@@ -0,0 +1,1485 @@
+// SPDX-License-Identifier: GPL-2.0
+/*
+ * Copyright (c) 2022, Teguh Sobirin.
+ * Copyright (c) 2024, Anton Bambura.
+ */
+
+/dts-v1/;
+
+#include <dt-bindings/leds/common.h>
+#include <dt-bindings/pinctrl/qcom,pmic-gpio.h>
+#include <dt-bindings/regulator/qcom,rpmh-regulator.h>
+#include <dt-bindings/sound/qcom,q6afe.h>
+#include <dt-bindings/sound/qcom,q6asm.h>
+
+#include "sdm845.dtsi"
+#include "sdm845-wcd9340.dtsi"
+#include "pm8998.dtsi"
+#include "pmi8998.dtsi"
+
+/ {
+	model = "AYN Odin";
+	compatible = "ayn,odin", "qcom,sdm845";
+	qcom,msm-id = <321 0x20001>;
+	qcom,board-id = <8 0>;
+	chassis-type = "tablet";
+
+	aliases {
+		serial0 = &uart9;
+		hsuart0 = &uart6;
+	};
+
+	chosen {
+		stdout-path = "serial0:115200n8";
+	};
+
+	gpio-keys {
+		compatible = "gpio-keys";
+		autorepeat;
+
+		pinctrl-names = "default";
+		pinctrl-0 = <&volume_up_gpio>;
+
+		key-volume-up {
+			label = "Volume Up";
+			linux,code = <KEY_VOLUMEUP>;
+			gpios = <&pm8998_gpios 6 GPIO_ACTIVE_LOW>;
+		};
+	};
+
+	led-controller {
+		compatible = "gpio-leds";
+
+		pinctrl-names = "default";
+		pinctrl-0 = <&side_leds_active>;
+
+		led-left {
+			color = <LED_COLOR_ID_BLUE>;
+			function = LED_FUNCTION_BACKLIGHT;
+			function-enumerator = <1>;
+			gpios = <&tlmm 129 GPIO_ACTIVE_HIGH>;
+			default-state = "off";
+		};
+
+		led-right {
+			color = <LED_COLOR_ID_BLUE>;
+			function = LED_FUNCTION_BACKLIGHT;
+			function-enumerator = <2>;
+			gpios = <&tlmm 87 GPIO_ACTIVE_HIGH>;
+			default-state = "off";
+		};
+
+		led-left-stick {
+			color = <LED_COLOR_ID_BLUE>;
+			function = LED_FUNCTION_BACKLIGHT;
+			function-enumerator = <3>;
+			gpios = <&tlmm 25 GPIO_ACTIVE_HIGH>;
+			default-state = "off";
+		};
+
+		led-right-stick {
+			color = <LED_COLOR_ID_BLUE>;
+			function = LED_FUNCTION_BACKLIGHT;
+			function-enumerator = <4>;
+			gpios = <&tlmm 7 GPIO_ACTIVE_HIGH>;
+			default-state = "off";
+		};
+	};
+
+	adc-joystick {
+		compatible = "adc-joystick";
+		io-channels = <&pm8998_adc ADC5_GPIO3>,
+			      <&pm8998_adc ADC5_GPIO2>,
+			      <&pm8998_adc ADC5_GPIO1>,
+			      <&pm8998_adc ADC5_GPIO5>,
+			      <&pm8998_adc ADC5_GPIO6_DIV3>,
+			      <&pm8998_adc ADC5_GPIO4>;
+		poll-interval = <10>;
+		#address-cells = <1>;
+		#size-cells = <0>;
+
+		axis@0 {
+			reg = <0>;
+			linux,code = <ABS_X>;
+			/* Stick neutral is near 850 mV, not ADC mid-scale. */
+			abs-range = <250000 1450000>;
+			abs-fuzz = <3000>;
+			abs-flat = <20000>;
+		};
+
+		axis@1 {
+			reg = <1>;
+			linux,code = <ABS_Y>;
+			abs-range = <1450000 250000>;
+			abs-fuzz = <3000>;
+			abs-flat = <20000>;
+		};
+
+		axis@2 {
+			reg = <2>;
+			linux,code = <ABS_Z>;
+			abs-range = <1425000 825000>;
+			abs-fuzz = <3000>;
+			abs-flat = <90000>;
+		};
+
+		axis@3 {
+			reg = <3>;
+			linux,code = <ABS_RX>;
+			abs-range = <1450000 250000>;
+			abs-fuzz = <3000>;
+			abs-flat = <20000>;
+		};
+
+		axis@4 {
+			reg = <4>;
+			linux,code = <ABS_RY>;
+			abs-range = <1450000 250000>;
+			abs-fuzz = <3000>;
+			abs-flat = <20000>;
+		};
+
+		axis@5 {
+			reg = <5>;
+			linux,code = <ABS_RZ>;
+			abs-range = <1425000 825000>;
+			abs-fuzz = <3000>;
+			abs-flat = <90000>;
+		};
+	};
+
+	gpio-keys-gamepad {
+		compatible = "gpio-keys";
+		label = "AYN Odin Gamepad Buttons";
+
+		pinctrl-names = "default";
+		pinctrl-0 = <&btn_pins &btn_en_active>;
+
+		button-east {
+			gpios = <&tlmm 96 GPIO_ACTIVE_LOW>;
+			label = "East";
+			linux,code = <BTN_EAST>;
+		};
+
+		button-south {
+			gpios = <&tlmm 95 GPIO_ACTIVE_LOW>;
+			label = "South";
+			linux,code = <BTN_SOUTH>;
+		};
+
+		button-west {
+			gpios = <&tlmm 41 GPIO_ACTIVE_LOW>;
+			label = "West";
+			linux,code = <BTN_WEST>;
+		};
+
+		button-north {
+			gpios = <&tlmm 92 GPIO_ACTIVE_LOW>;
+			label = "North";
+			linux,code = <BTN_NORTH>;
+		};
+
+		button-left-shoulder {
+			gpios = <&tlmm 26 GPIO_ACTIVE_LOW>;
+			label = "Left Shoulder";
+			linux,code = <BTN_TL>;
+		};
+
+		button-right-shoulder {
+			gpios = <&tlmm 91 GPIO_ACTIVE_LOW>;
+			label = "Right Shoulder";
+			linux,code = <BTN_TR>;
+		};
+
+		button-select {
+			gpios = <&tlmm 0 GPIO_ACTIVE_LOW>;
+			label = "Select";
+			linux,code = <BTN_SELECT>;
+		};
+
+		button-start {
+			gpios = <&tlmm 3 GPIO_ACTIVE_LOW>;
+			label = "Start";
+			linux,code = <BTN_START>;
+		};
+
+		button-left-stick {
+			gpios = <&tlmm 40 GPIO_ACTIVE_LOW>;
+			label = "Left Stick";
+			linux,code = <BTN_THUMBL>;
+		};
+
+		button-right-stick {
+			gpios = <&tlmm 22 GPIO_ACTIVE_LOW>;
+			label = "Right Stick";
+			linux,code = <BTN_THUMBR>;
+		};
+
+		button-dpad-up {
+			gpios = <&tlmm 124 GPIO_ACTIVE_LOW>;
+			label = "D-pad Up";
+			linux,code = <BTN_DPAD_UP>;
+		};
+
+		button-dpad-down {
+			gpios = <&tlmm 128 GPIO_ACTIVE_LOW>;
+			label = "D-pad Down";
+			linux,code = <BTN_DPAD_DOWN>;
+		};
+
+		button-dpad-left {
+			gpios = <&tlmm 52 GPIO_ACTIVE_LOW>;
+			label = "D-pad Left";
+			linux,code = <BTN_DPAD_LEFT>;
+		};
+
+		button-dpad-right {
+			gpios = <&tlmm 122 GPIO_ACTIVE_LOW>;
+			label = "D-pad Right";
+			linux,code = <BTN_DPAD_RIGHT>;
+		};
+
+		button-mode {
+			gpios = <&tlmm 49 GPIO_ACTIVE_LOW>;
+			label = "Mode";
+			linux,code = <BTN_MODE>;
+		};
+
+		button-right-paddle {
+			gpios = <&tlmm 24 GPIO_ACTIVE_LOW>;
+			label = "Right Paddle";
+			linux,code = <BTN_TRIGGER_HAPPY5>;
+		};
+
+		button-left-paddle {
+			gpios = <&tlmm 116 GPIO_ACTIVE_LOW>;
+			label = "Left Paddle";
+			linux,code = <BTN_TRIGGER_HAPPY7>;
+		};
+	};
+
+	battery: battery {
+		compatible = "simple-battery";
+
+		charge-full-design-microamp-hours = <3850000>;
+		voltage-min-design-microvolt = <3600000>;
+		voltage-max-design-microvolt = <4400000>;
+	};
+
+	fan: pwm-fan {
+		compatible = "pwm-fan";
+		#cooling-cells = <2>;
+
+		pwms = <&pmi8998_lpg 1 200000>;
+		fan-supply = <&fan_pwr>;
+		interrupts-extended = <&tlmm 30 IRQ_TYPE_EDGE_RISING>;
+		pulses-per-revolution = <2>;
+
+		pwm-inverted;
+		initial-pwm = <0>;
+		fan-supply-off-on-zero;
+		cooling-levels = <0 80 96 128 160 200 255>;
+
+		pinctrl-names = "default";
+		pinctrl-0 = <&fan_tach_active &pwm_fan_active>;
+	};
+
+	connector {
+		compatible = "hdmi-connector";
+
+		label = "hdmi";
+		type = "d";
+
+		hdmi-pwr-supply = <&hdmi_en_reg>;
+		hpd-gpios = <&tlmm 121 GPIO_ACTIVE_HIGH>;
+
+		pinctrl-names = "default";
+		pinctrl-0 = <&hdmi_hpd>;
+
+		port {
+			hdmi_con: endpoint {
+				remote-endpoint = <&lt8912b_out>;
+			};
+		};
+	};
+
+	hdmi_en_reg: hdmi-en-regulator {
+		compatible = "regulator-fixed";
+		regulator-name = "hdmi_en";
+
+		gpio = <&tlmm 13 GPIO_ACTIVE_HIGH>;
+		enable-active-high;
+	};
+
+	fan_pwr: fan-pwr-regulator {
+		compatible = "regulator-fixed";
+		regulator-name = "fan_pwr";
+
+		gpio = <&tlmm 44 GPIO_ACTIVE_HIGH>;
+		enable-active-high;
+
+		pinctrl-names = "default";
+		pinctrl-0 = <&fan_en_active>;
+	};
+
+	vph_pwr: vph-pwr-regulator {
+		compatible = "regulator-fixed";
+		regulator-name = "vph_pwr";
+		regulator-min-microvolt = <3700000>;
+		regulator-max-microvolt = <3700000>;
+	};
+
+	vreg_s4a_1p8: pm8998-smps4 {
+		compatible = "regulator-fixed";
+		regulator-name = "vreg_s4a_1p8";
+
+		regulator-min-microvolt = <1800000>;
+		regulator-max-microvolt = <1800000>;
+
+		regulator-always-on;
+		regulator-boot-on;
+
+		vin-supply = <&vph_pwr>;
+	};
+};
+
+&cpu0_alert0 {
+	temperature = <78000>;
+	hysteresis = <5000>;
+	type = "active";
+};
+
+&cpu0_alert1 {
+	temperature = <82000>;
+	hysteresis = <5000>;
+	type = "active";
+};
+
+&cpu1_alert0 {
+	temperature = <78000>;
+	hysteresis = <5000>;
+	type = "active";
+};
+
+&cpu1_alert1 {
+	temperature = <82000>;
+	hysteresis = <5000>;
+	type = "active";
+};
+
+&cpu2_alert0 {
+	temperature = <78000>;
+	hysteresis = <5000>;
+	type = "active";
+};
+
+&cpu2_alert1 {
+	temperature = <82000>;
+	hysteresis = <5000>;
+	type = "active";
+};
+
+&cpu3_alert0 {
+	temperature = <78000>;
+	hysteresis = <5000>;
+	type = "active";
+};
+
+&cpu3_alert1 {
+	temperature = <82000>;
+	hysteresis = <5000>;
+	type = "active";
+};
+
+&cpu4_alert0 {
+	temperature = <78000>;
+	hysteresis = <5000>;
+	type = "active";
+};
+
+&cpu4_alert1 {
+	temperature = <82000>;
+	hysteresis = <5000>;
+	type = "active";
+};
+
+&cpu5_alert0 {
+	temperature = <78000>;
+	hysteresis = <5000>;
+	type = "active";
+};
+
+&cpu5_alert1 {
+	temperature = <82000>;
+	hysteresis = <5000>;
+	type = "active";
+};
+
+&cpu6_alert0 {
+	temperature = <78000>;
+	hysteresis = <5000>;
+	type = "active";
+};
+
+&cpu6_alert1 {
+	temperature = <82000>;
+	hysteresis = <5000>;
+	type = "active";
+};
+
+&cpu7_alert0 {
+	temperature = <78000>;
+	hysteresis = <5000>;
+	type = "active";
+};
+
+&cpu7_alert1 {
+	temperature = <82000>;
+	hysteresis = <5000>;
+	type = "active";
+};
+
+/*
+ * Keep the fan off for idle and light loads, then use active cooling before
+ * performance throttling: Fan trips start quietly at 78 C and ramp harder at 82 C.
+ */
+&cpu0_thermal {
+	cooling-maps {
+		map0 {
+			trip = <&cpu0_alert0>;
+			cooling-device = <&fan 0 1>;
+		};
+
+		map1 {
+			trip = <&cpu0_alert1>;
+			cooling-device = <&fan 0 3>;
+		};
+	};
+};
+
+&cpu1_thermal {
+	cooling-maps {
+		map0 {
+			trip = <&cpu1_alert0>;
+			cooling-device = <&fan 0 1>;
+		};
+
+		map1 {
+			trip = <&cpu1_alert1>;
+			cooling-device = <&fan 0 3>;
+		};
+	};
+};
+
+&cpu2_thermal {
+	cooling-maps {
+		map0 {
+			trip = <&cpu2_alert0>;
+			cooling-device = <&fan 0 1>;
+		};
+
+		map1 {
+			trip = <&cpu2_alert1>;
+			cooling-device = <&fan 0 3>;
+		};
+	};
+};
+
+&cpu3_thermal {
+	cooling-maps {
+		map0 {
+			trip = <&cpu3_alert0>;
+			cooling-device = <&fan 0 1>;
+		};
+
+		map1 {
+			trip = <&cpu3_alert1>;
+			cooling-device = <&fan 0 3>;
+		};
+	};
+};
+
+&cpu4_thermal {
+	cooling-maps {
+		map0 {
+			trip = <&cpu4_alert0>;
+			cooling-device = <&fan 0 1>;
+		};
+
+		map1 {
+			trip = <&cpu4_alert1>;
+			cooling-device = <&fan 0 3>;
+		};
+	};
+};
+
+&cpu5_thermal {
+	cooling-maps {
+		map0 {
+			trip = <&cpu5_alert0>;
+			cooling-device = <&fan 0 1>;
+		};
+
+		map1 {
+			trip = <&cpu5_alert1>;
+			cooling-device = <&fan 0 3>;
+		};
+	};
+};
+
+&cpu6_thermal {
+	cooling-maps {
+		map0 {
+			trip = <&cpu6_alert0>;
+			cooling-device = <&fan 0 1>;
+		};
+
+		map1 {
+			trip = <&cpu6_alert1>;
+			cooling-device = <&fan 0 3>;
+		};
+	};
+};
+
+&cpu7_thermal {
+	cooling-maps {
+		map0 {
+			trip = <&cpu7_alert0>;
+			cooling-device = <&fan 0 1>;
+		};
+
+		map1 {
+			trip = <&cpu7_alert1>;
+			cooling-device = <&fan 0 3>;
+		};
+	};
+};
+
+&gpu_top_thermal {
+	trips {
+		gpu_top_fan_alert0: fan-alert0 {
+			temperature = <78000>;
+			hysteresis = <5000>;
+			type = "active";
+		};
+
+		gpu_top_fan_alert1: fan-alert1 {
+			temperature = <82000>;
+			hysteresis = <5000>;
+			type = "active";
+		};
+	};
+
+	cooling-maps {
+		fan-map0 {
+			trip = <&gpu_top_fan_alert0>;
+			cooling-device = <&fan 0 1>;
+		};
+
+		fan-map1 {
+			trip = <&gpu_top_fan_alert1>;
+			cooling-device = <&fan 0 3>;
+		};
+	};
+};
+
+&gpu_bottom_thermal {
+	trips {
+		gpu_bottom_fan_alert0: fan-alert0 {
+			temperature = <78000>;
+			hysteresis = <5000>;
+			type = "active";
+		};
+
+		gpu_bottom_fan_alert1: fan-alert1 {
+			temperature = <82000>;
+			hysteresis = <5000>;
+			type = "active";
+		};
+	};
+
+	cooling-maps {
+		fan-map0 {
+			trip = <&gpu_bottom_fan_alert0>;
+			cooling-device = <&fan 0 1>;
+		};
+
+		fan-map1 {
+			trip = <&gpu_bottom_fan_alert1>;
+			cooling-device = <&fan 0 3>;
+		};
+	};
+};
+
+&adsp_pas {
+	status = "okay";
+	firmware-name = "qcom/sdm845/adsp.mbn";
+};
+
+&apps_rsc {
+	pm8998-rpmh-regulators {
+		compatible = "qcom,pm8998-rpmh-regulators";
+		qcom,pmic-id = "a";
+
+		vdd-s1-supply = <&vph_pwr>;
+		vdd-s2-supply = <&vph_pwr>;
+		vdd-s3-supply = <&vph_pwr>;
+		vdd-s4-supply = <&vph_pwr>;
+		vdd-s5-supply = <&vph_pwr>;
+		vdd-s6-supply = <&vph_pwr>;
+		vdd-s7-supply = <&vph_pwr>;
+		vdd-s8-supply = <&vph_pwr>;
+		vdd-s9-supply = <&vph_pwr>;
+		vdd-s10-supply = <&vph_pwr>;
+		vdd-s11-supply = <&vph_pwr>;
+		vdd-s12-supply = <&vph_pwr>;
+		vdd-s13-supply = <&vph_pwr>;
+		vdd-l1-l27-supply = <&vreg_s7a_1p025>;
+		vdd-l2-l8-l17-supply = <&vreg_s3a_1p35>;
+		vdd-l3-l11-supply = <&vreg_s7a_1p025>;
+		vdd-l4-l5-supply = <&vreg_s7a_1p025>;
+		vdd-l6-supply = <&vph_pwr>;
+		vdd-l7-l12-l14-l15-supply = <&vreg_s5a_2p04>;
+		vdd-l9-supply = <&vreg_bob>;
+		vdd-l10-l23-l25-supply = <&vreg_bob>;
+		vdd-l13-l19-l21-supply = <&vreg_bob>;
+		vdd-l16-l28-supply = <&vreg_bob>;
+		vdd-l18-l22-supply = <&vreg_bob>;
+		vdd-l20-l24-supply = <&vreg_bob>;
+		vdd-l26-supply = <&vreg_s3a_1p35>;
+		vin-lvs-1-2-supply = <&vreg_s4a_1p8>;
+
+		vreg_s3a_1p35: smps3 {
+			regulator-min-microvolt = <1352000>;
+			regulator-max-microvolt = <1352000>;
+		};
+
+		vreg_s5a_2p04: smps5 {
+			regulator-min-microvolt = <1904000>;
+			regulator-max-microvolt = <2040000>;
+		};
+
+		vreg_s7a_1p025: smps7 {
+			regulator-min-microvolt = <900000>;
+			regulator-max-microvolt = <1028000>;
+		};
+
+		vdda_mipi_dsi0_pll:
+		vdda_mipi_dsi1_pll:
+		vdda_qlink_lv:
+		vdda_ufs1_core:
+		vdda_usb1_ss_core:
+		vdda_qusb_hs0_1p8:
+		vdda_qusb_hs0_3p1:
+		vreg_l1a_0p875: ldo1 {
+			regulator-min-microvolt = <880000>;
+			regulator-max-microvolt = <880000>;
+			regulator-initial-mode = <RPMH_REGULATOR_MODE_HPM>;
+		};
+
+		vreg_l5a_0p8: ldo5 {
+			regulator-min-microvolt = <800000>;
+			regulator-max-microvolt = <800000>;
+			regulator-initial-mode = <RPMH_REGULATOR_MODE_HPM>;
+		};
+
+		vreg_l7a_1p8: ldo7 {
+			regulator-min-microvolt = <1800000>;
+			regulator-max-microvolt = <1800000>;
+			regulator-initial-mode = <RPMH_REGULATOR_MODE_HPM>;
+		};
+
+		vreg_l12a_1p8: ldo12 {
+			regulator-min-microvolt = <1800000>;
+			regulator-max-microvolt = <1800000>;
+			regulator-initial-mode = <RPMH_REGULATOR_MODE_HPM>;
+		};
+
+		vreg_l13a_2p95: ldo13 {
+			regulator-min-microvolt = <1800000>;
+			regulator-max-microvolt = <2960000>;
+			regulator-initial-mode = <RPMH_REGULATOR_MODE_HPM>;
+		};
+
+		vreg_l14a_1p8: ldo14 {
+			regulator-min-microvolt = <1800000>;
+			regulator-max-microvolt = <1800000>;
+			regulator-initial-mode = <RPMH_REGULATOR_MODE_HPM>;
+			regulator-boot-on;
+			regulator-always-on;
+		};
+
+		vreg_l17a_1p3: ldo17 {
+			regulator-min-microvolt = <1304000>;
+			regulator-max-microvolt = <1304000>;
+			regulator-initial-mode = <RPMH_REGULATOR_MODE_HPM>;
+		};
+
+		vreg_l20a_2p95: ldo20 {
+			regulator-min-microvolt = <2950000>;
+			regulator-max-microvolt = <2960000>;
+			regulator-initial-mode = <RPMH_REGULATOR_MODE_HPM>;
+		};
+
+		vreg_l21a_2p95: ldo21 {
+			regulator-min-microvolt = <2950000>;
+			regulator-max-microvolt = <2960000>;
+			regulator-initial-mode = <RPMH_REGULATOR_MODE_HPM>;
+		};
+
+		vreg_l23a_3p3: ldo23 {
+			regulator-min-microvolt = <3300000>;
+			regulator-max-microvolt = <3312000>;
+			regulator-initial-mode = <RPMH_REGULATOR_MODE_HPM>;
+		};
+
+		vreg_l24a_3p075: ldo24 {
+			regulator-min-microvolt = <3088000>;
+			regulator-max-microvolt = <3088000>;
+			regulator-initial-mode = <RPMH_REGULATOR_MODE_HPM>;
+		};
+
+		vreg_l25a_3p3: ldo25 {
+			regulator-min-microvolt = <3300000>;
+			regulator-max-microvolt = <3312000>;
+			regulator-initial-mode = <RPMH_REGULATOR_MODE_HPM>;
+		};
+
+		vreg_l26a_1p2: ldo26 {
+			regulator-min-microvolt = <1200000>;
+			regulator-max-microvolt = <1200000>;
+			regulator-initial-mode = <RPMH_REGULATOR_MODE_HPM>;
+			regulator-boot-on;
+		};
+
+		vreg_l28a_3p0: ldo28 {
+			regulator-min-microvolt = <2856000>;
+			regulator-max-microvolt = <3008000>;
+			regulator-initial-mode = <RPMH_REGULATOR_MODE_HPM>;
+		};
+	};
+
+	pmi8998-rpmh-regulators {
+		compatible = "qcom,pmi8998-rpmh-regulators";
+		qcom,pmic-id = "b";
+
+		vdd-bob-supply = <&vph_pwr>;
+
+		vreg_bob: bob {
+			regulator-min-microvolt = <3312000>;
+			regulator-max-microvolt = <3600000>;
+			regulator-initial-mode = <RPMH_REGULATOR_MODE_AUTO>;
+			regulator-allow-bypass;
+		};
+	};
+
+	pm8005-rpmh-regulators {
+		compatible = "qcom,pm8005-rpmh-regulators";
+		qcom,pmic-id = "c";
+
+		vdd-s1-supply = <&vph_pwr>;
+		vdd-s2-supply = <&vph_pwr>;
+		vdd-s3-supply = <&vph_pwr>;
+		vdd-s4-supply = <&vph_pwr>;
+
+		vreg_s3c_0p6: smps3 {
+			regulator-min-microvolt = <600000>;
+			regulator-max-microvolt = <600000>;
+		};
+	};
+};
+
+&cdsp_pas {
+	status = "okay";
+	firmware-name = "qcom/sdm845/cdsp.mbn";
+};
+
+&mdss {
+	status = "okay";
+};
+
+&mdss_dsi0 {
+	status = "okay";
+	vdda-supply = <&vreg_l26a_1p2>;
+
+	display_panel: panel@0 {
+		status = "disabled";
+		reg = <0>;
+		vddio-supply = <&vreg_l14a_1p8>;
+		vddpos-supply = <&lab>;
+		vddneg-supply = <&ibb>;
+
+		backlight = <&pmi8998_wled>;
+		enable-gpios = <&tlmm 120 GPIO_ACTIVE_HIGH>;
+		reset-gpios = <&tlmm 6 GPIO_ACTIVE_LOW>;
+		te-gpios = <&tlmm 10 GPIO_ACTIVE_HIGH>;
+
+		pinctrl-names = "default";
+		pinctrl-0 = <&panel_enable_active &panel_reset_active
+			     &panel_te_active>;
+
+		port {
+			panel_in: endpoint {
+				remote-endpoint = <&mdss_dsi0_out>;
+			};
+		};
+	};
+};
+
+&mdss_dsi0_out {
+	remote-endpoint = <&panel_in>;
+	data-lanes = <0 1 2 3>;
+};
+
+&mdss_dsi0_phy {
+	status = "okay";
+	vdds-supply = <&vdda_mipi_dsi0_pll>;
+};
+
+&mdss_dsi1 {
+	vdda-supply = <&vreg_l26a_1p2>;
+	status = "okay";
+};
+
+&mdss_dsi1_out {
+	remote-endpoint = <&lt8912b_in>;
+	data-lanes = <0 1 2 3>;
+};
+
+&mdss_dsi1_phy {
+	vdds-supply = <&vreg_l1a_0p875>;
+	status = "okay";
+};
+
+&dispcc {
+	status = "okay";
+};
+
+&gcc {
+	protected-clocks = <GCC_QSPI_CORE_CLK>,
+			   <GCC_QSPI_CORE_CLK_SRC>,
+			   <GCC_QSPI_CNOC_PERIPH_AHB_CLK>,
+			   <GCC_LPASS_Q6_AXI_CLK>,
+			   <GCC_LPASS_SWAY_CLK>;
+};
+
+&gmu {
+	status = "okay";
+};
+
+&gpi_dma0 {
+	status = "okay";
+};
+
+&gpi_dma1 {
+	status = "okay";
+};
+
+&gpu {
+	status = "okay";
+};
+
+&gpu_zap_shader {
+	firmware-name = "qcom/sdm845/a630_zap.mbn";
+};
+
+&i2c4 {
+	status = "okay";
+	clock-frequency = <400000>;
+
+	hdmi-bridge@48 {
+		compatible = "lontium,lt8912b";
+		reg = <0x48>;
+
+		// vdd-supply = <&vreg_s4a_1p8>;
+		// vccmipirx-supply = <&vreg_s4a_1p8>;
+		// vccsysclk-supply = <&vreg_s4a_1p8>;
+		// vcclvdstx-supply = <&vreg_s4a_1p8>;
+		// vcchdmitx-supply = <&vreg_s4a_1p8>;
+		// vcclvdspll-supply = <&vreg_s4a_1p8>;
+		// vcchdmipll-supply = <&vreg_s4a_1p8>;
+
+		reset-gpios = <&tlmm 12 GPIO_ACTIVE_LOW>;
+
+		ports {
+			#address-cells = <1>;
+			#size-cells = <0>;
+
+			port@0 {
+				reg = <0>;
+
+				lt8912b_in: endpoint {
+					data-lanes = <0 1 2 3>;
+					remote-endpoint = <&mdss_dsi1_out>;
+				};
+			};
+
+			port@1 {
+				reg = <1>;
+
+				lt8912b_out: endpoint {
+					remote-endpoint = <&hdmi_con>;
+				};
+			};
+		};
+	};
+};
+
+&i2c5 {
+	status = "okay";
+	clock-frequency = <400000>;
+};
+
+&swm {
+	left_spkr: wsa8810-left {
+		compatible = "sdw10217201000";
+		reg = <0 1>;
+		powerdown-gpios = <&wcdgpio 2 GPIO_ACTIVE_HIGH>;
+		#thermal-sensor-cells = <0>;
+		sound-name-prefix = "SpkrLeft";
+		#sound-dai-cells = <0>;
+	};
+
+	right_spkr: wsa8810-right {
+		compatible = "sdw10217201000";
+		powerdown-gpios = <&wcdgpio 2 GPIO_ACTIVE_HIGH>;
+		reg = <0 2>;
+		#thermal-sensor-cells = <0>;
+		sound-name-prefix = "SpkrRight";
+		#sound-dai-cells = <0>;
+	};
+};
+
+&ibb {
+	regulator-min-microvolt = <5500000>;
+	regulator-max-microvolt = <5500000>;
+	regulator-pull-down;
+	regulator-soft-start;
+	qcom,discharge-resistor-kohms = <300>;
+};
+
+&lab {
+	regulator-min-microvolt = <5500000>;
+	regulator-max-microvolt = <5500000>;
+	regulator-pull-down;
+	regulator-soft-start;
+	qcom,soft-start-us = <800>;
+};
+
+&mss_pil {
+	status = "okay";
+	firmware-name = "qcom/sdm845/mba.mbn", "qcom/sdm845/modem.mbn";
+};
+
+&pm8998_adc {
+	adc-chan@12 {
+		reg = <ADC5_GPIO1>;
+		label = "gpio1";
+		qcom,decimation = <256>;
+		qcom,pre-scaling = <1 1>;
+	};
+
+	adc-chan@13 {
+		reg = <ADC5_GPIO2>;
+		label = "gpio2";
+		qcom,decimation = <256>;
+		qcom,pre-scaling = <1 1>;
+	};
+
+	adc-chan@14 {
+		reg = <ADC5_GPIO3>;
+		label = "gpio3";
+		qcom,decimation = <256>;
+		qcom,pre-scaling = <1 1>;
+	};
+
+	adc-chan@15 {
+		reg = <ADC5_GPIO4>;
+		label = "gpio4";
+		qcom,decimation = <256>;
+		qcom,pre-scaling = <1 1>;
+	};
+
+	adc-chan@16 {
+		reg = <ADC5_GPIO5>;
+		label = "gpio5";
+		qcom,decimation = <256>;
+		qcom,pre-scaling = <1 1>;
+	};
+
+	adc-chan@97 {
+		reg = <ADC5_GPIO6_DIV3>;
+		label = "gpio6";
+		qcom,decimation = <256>;
+		qcom,pre-scaling = <1 3>;
+	};
+};
+
+&pm8998_gpios {
+	volume_up_gpio: vol-up-active {
+		pins = "gpio6";
+		function = "normal";
+		input-enable;
+		bias-pull-up;
+		qcom,drive-strength = <PMIC_GPIO_STRENGTH_NO>;
+	};
+};
+
+&pmi8998_gpios {
+	pwm_fan_active: pwm-fan-active {
+		pins = "gpio5";
+		function = "func1";
+		bias-disable;
+		power-source = <0>;
+		output-high;
+		qcom,drive-strength = <PMIC_GPIO_STRENGTH_LOW>;
+		drive-push-pull;
+	};
+};
+
+&pmi8998_lpg {
+	status = "okay";
+
+	multi-led {
+		color = <LED_COLOR_ID_RGB>;
+		function = LED_FUNCTION_STATUS;
+
+		#address-cells = <1>;
+		#size-cells = <0>;
+
+		led@3 {
+			reg = <3>;
+			color = <LED_COLOR_ID_BLUE>;
+		};
+
+		led@4 {
+			reg = <4>;
+			color = <LED_COLOR_ID_GREEN>;
+		};
+
+		led@5 {
+			reg = <5>;
+			color = <LED_COLOR_ID_RED>;
+		};
+	};
+};
+
+&pmi8998_haptics {
+	qcom,actuator-type = <HAP_TYPE_ERM>;
+	qcom,play-mode = <HAP_PLAY_DIRECT>;
+	qcom,wave-shape = <HAP_WAVE_SQUARE>;
+	qcom,wave-play-rate-us = <6667>;
+
+	status = "okay";
+};
+
+&pm8998_resin {
+	linux,code = <KEY_VOLUMEDOWN>;
+	status = "okay";
+};
+
+&pmi8998_charger {
+	status = "okay";
+	monitored-battery = <&battery>;
+};
+
+&pmi8998_fg {
+	status = "okay";
+	monitored-battery = <&battery>;
+	power-supplies = <&pmi8998_charger>;
+};
+
+&pmi8998_rradc {
+	status = "okay";
+	monitored-battery = <&battery>;
+};
+
+&pmi8998_wled {
+	status = "okay";
+	qcom,enabled-strings = <0 1>;
+	qcom,num-strings = <2>;
+};
+
+&q6asmdai {
+	dai@0 {
+		reg = <0>;
+	};
+
+	dai@1 {
+		reg = <1>;
+	};
+
+	dai@2 {
+		reg = <2>;
+	};
+};
+
+&qupv3_id_0 {
+	status = "okay";
+};
+
+&qupv3_id_1 {
+	status = "okay";
+};
+
+&qup_uart6_default {
+	pinmux {
+		pins = "gpio45", "gpio46", "gpio47", "gpio48";
+		function = "qup6";
+	};
+
+	cts {
+		pins = "gpio45";
+		bias-disable;
+	};
+
+	rts-tx {
+		pins = "gpio46", "gpio47";
+		drive-strength = <2>;
+		bias-disable;
+	};
+
+	rx {
+		pins = "gpio48";
+		bias-pull-up;
+	};
+};
+
+&qup_uart9_default {
+	pinconf-tx {
+		pins = "gpio4";
+		drive-strength = <2>;
+		bias-disable;
+	};
+
+	pinconf-rx {
+		pins = "gpio5";
+		drive-strength = <2>;
+		bias-pull-up;
+	};
+};
+
+&sdhc_2 {
+	status = "okay";
+
+	pinctrl-names = "default";
+	pinctrl-0 = <&sdc2_default_state &sdc2_card_det_n>;
+
+	vmmc-supply = <&vreg_l21a_2p95>;
+	vqmmc-supply = <&vreg_l13a_2p95>;
+
+	bus-width = <4>;
+	cd-gpios = <&tlmm 126 GPIO_ACTIVE_LOW>;
+};
+
+&sound {
+	compatible = "ayn,odin-sndcard", "qcom,sdm845-sndcard";
+	model = "AYN Odin";
+
+	audio-routing =
+		"RX_BIAS", "MCLK",
+		"AMIC1", "MIC BIAS1",
+		"AMIC2", "MIC BIAS2",
+		"SpkrLeft IN", "SPK1 OUT",
+		"SpkrRight IN", "SPK2 OUT",
+		"MM_DL1",  "MultiMedia1 Playback",
+		"MM_DL3",  "MultiMedia3 Playback",
+		"MultiMedia2 Capture", "MM_UL2";
+
+	mm1-dai-link {
+		link-name = "MultiMedia1";
+		cpu {
+			sound-dai = <&q6asmdai  MSM_FRONTEND_DAI_MULTIMEDIA1>;
+		};
+	};
+
+	mm2-dai-link {
+		link-name = "MultiMedia2";
+		cpu {
+			sound-dai = <&q6asmdai  MSM_FRONTEND_DAI_MULTIMEDIA2>;
+		};
+	};
+
+	mm3-dai-link {
+		link-name = "MultiMedia3";
+		cpu {
+			sound-dai = <&q6asmdai  MSM_FRONTEND_DAI_MULTIMEDIA3>;
+		};
+	};
+
+	slim-dai-link {
+		link-name = "SLIM Playback";
+		cpu {
+			sound-dai = <&q6afedai SLIMBUS_0_RX>;
+		};
+
+		platform {
+			sound-dai = <&q6routing>;
+		};
+
+		codec {
+			sound-dai =  <&left_spkr>, <&right_spkr>, <&swm 0>, <&wcd9340 0>;
+		};
+	};
+
+	slimcap-dai-link {
+		link-name = "SLIM Capture";
+		cpu {
+			sound-dai = <&q6afedai SLIMBUS_0_TX>;
+		};
+
+		platform {
+			sound-dai = <&q6routing>;
+		};
+
+		codec {
+			sound-dai = <&wcd9340 1>;
+		};
+	};
+
+	slim-wcd-dai-link {
+		link-name = "SLIM WCD Playback";
+		cpu {
+			sound-dai = <&q6afedai SLIMBUS_1_RX>;
+		};
+
+		platform {
+			sound-dai = <&q6routing>;
+		};
+
+		codec {
+			sound-dai =  <&wcd9340 2>;
+		};
+	};
+};
+
+&slpi_pas {
+	firmware-name = "qcom/sdm845/AYN/Odin/slpi.mbn";
+	status = "okay";
+};
+
+&tlmm {
+	sdc2_default_state: sdc2-default {
+		clk {
+			pins = "sdc2_clk";
+			bias-disable;
+			drive-strength = <16>;
+		};
+
+		cmd {
+			pins = "sdc2_cmd";
+			bias-pull-up;
+			drive-strength = <10>;
+		};
+
+		data {
+			pins = "sdc2_data";
+			bias-pull-up;
+			drive-strength = <10>;
+		};
+	};
+
+	sdc2_card_det_n: sd-card-det-n {
+		pins = "gpio126";
+		function = "gpio";
+		bias-pull-up;
+	};
+
+	ts_int_active: ts-int-active {
+		pins = "gpio125";
+		function = "gpio";
+		drive-strength = <16>;
+		bias-pull-up;
+		input-enable;
+	};
+
+	ts_reset_active: ts-reset-active {
+		pins = "gpio104";
+		function = "gpio";
+		drive-strength = <16>;
+		bias-pull-up;
+	};
+
+	btn_en_active: btn-en-active {
+		pins = "gpio88";
+		function = "gpio";
+		drive-strength = <8>;
+		bias-disable;
+		output-high;
+	};
+
+	side_leds_active: side-leds-active {
+		pins = "gpio7", "gpio25", "gpio87", "gpio129";
+		function = "gpio";
+		drive-strength = <2>;
+		bias-disable;
+	};
+
+	btn_pins: btn-pins {
+		mux {
+			pins =	"gpio96", "gpio95", "gpio41", "gpio92",
+				"gpio26", "gpio91", "gpio0", "gpio3",
+				"gpio40", "gpio22", "gpio124", "gpio128",
+				"gpio52", "gpio122", "gpio49", "gpio24",
+				"gpio116";
+			function = "gpio";
+			drive-strength = <16>;
+			input-enable;
+			bias-pull-up;
+		};
+	};
+
+	fan_en_active: fan-en-active {
+		pins = "gpio44";
+		function = "gpio";
+		drive-strength = <2>;
+		bias-pull-down;
+	};
+
+	fan_tach_active: fan-tach-active {
+		pins = "gpio30";
+		function = "gpio";
+		drive-strength = <8>;
+		bias-disable;
+		input-enable;
+	};
+
+	panel_reset_active: panel-reset-active {
+		pins = "gpio6";
+		function = "gpio";
+		drive-strength = <8>;
+		bias-disable;
+	};
+
+	panel_enable_active: panel-enable-active {
+		pins = "gpio120";
+		function = "gpio";
+		drive-strength = <2>;
+		bias-disable;
+	};
+
+	panel_te_active: panel-te-active {
+		pins = "gpio10";
+		function = "mdp_vsync";
+		drive-strength = <2>;
+		bias-disable;
+	};
+
+	hdmi_hpd: hdmi-hpd-active {
+		pins = "gpio121";
+		function = "gpio";
+		drive-strength = <10>;
+		bias-pull-up;
+		input-enable;
+	};
+};
+
+&uart6 {
+	status = "okay";
+
+	bluetooth {
+		compatible = "qcom,wcn3990-bt";
+
+		vddio-supply = <&vreg_s4a_1p8>;
+		vddxo-supply = <&vreg_l7a_1p8>;
+		vddrf-supply = <&vreg_l17a_1p3>;
+		vddch0-supply = <&vreg_l25a_3p3>;
+		vddch1-supply = <&vreg_l23a_3p3>;
+		max-speed = <3200000>;
+	};
+};
+
+&uart9 {
+	status = "okay";
+};
+
+&ufs_mem_hc {
+	status = "okay";
+
+	reset-gpios = <&tlmm 150 GPIO_ACTIVE_LOW>;
+
+	vcc-supply = <&vreg_l20a_2p95>;
+	vcc-max-microamp = <600000>;
+};
+
+&ufs_mem_phy {
+	status = "okay";
+
+	vdda-phy-supply = <&vreg_l1a_0p875>;
+	vdda-pll-supply = <&vreg_l26a_1p2>;
+};
+
+&usb_1 {
+	status = "okay";
+};
+
+&usb_1_dwc3 {
+	dr_mode = "otg";
+	usb-role-switch;
+	role-switch-default-mode = "peripheral";
+};
+
+&usb_1_hsphy {
+	status = "okay";
+
+	vdd-supply = <&vreg_l1a_0p875>;
+	vdda-pll-supply = <&vreg_l12a_1p8>;
+	vdda-phy-dpdm-supply = <&vreg_l24a_3p075>;
+
+	qcom,imp-res-offset-value = <8>;
+	qcom,hstx-trim-value = <QUSB2_V2_HSTX_TRIM_21_6_MA>;
+	qcom,preemphasis-level = <QUSB2_V2_PREEMPHASIS_5_PERCENT>;
+	qcom,preemphasis-width = <QUSB2_V2_PREEMPHASIS_WIDTH_HALF_BIT>;
+};
+
+&usb_1_qmpphy {
+	status = "okay";
+
+	vdda-phy-supply = <&vreg_l26a_1p2>;
+	vdda-pll-supply = <&vreg_l1a_0p875>;
+};
+
+&venus {
+	status = "okay";
+	firmware-name = "qcom/venus-5.2/venus.mbn";
+};
+
+&wcd9340 {
+	reset-gpios = <&tlmm 64 GPIO_ACTIVE_HIGH>;
+	vdd-buck-supply = <&vreg_s4a_1p8>;
+	vdd-buck-sido-supply = <&vreg_s4a_1p8>;
+	vdd-tx-supply = <&vreg_s4a_1p8>;
+	vdd-rx-supply = <&vreg_s4a_1p8>;
+	vdd-io-supply = <&vreg_s4a_1p8>;
+	qcom,mbhc-buttons-vthreshold-microvolt = <75000 150000 237000 500000 500000 500000 500000 500000>;
+	qcom,mbhc-headset-vthreshold-microvolt = <1700000>;
+	qcom,mbhc-headphone-vthreshold-microvolt = <50000>;
+};
+
+&wifi {
+	status = "okay";
+
+	vdd-0.8-cx-mx-supply = <&vreg_l5a_0p8>;
+	vdd-1.8-xo-supply = <&vreg_l7a_1p8>;
+	vdd-1.3-rfa-supply = <&vreg_l17a_1p3>;
+	vdd-3.3-ch0-supply = <&vreg_l25a_3p3>;
+	vdd-3.3-ch1-supply = <&vreg_l23a_3p3>;
+
+	qcom,snoc-host-cap-8bit-quirk;
+};
+
+/* PINCTRL - additions to nodes defined in sdm845.dtsi */
+&qup_spi0_default {
+	drive-strength = <6>;
+	bias-disable;
+};
+
+&qup_i2c4_default {
+	drive-strength = <2>;
+	bias-disable;
+};
+
+&qup_i2c5_default {
+	drive-strength = <2>;
+	bias-disable;
+};
diff --git a/arch/arm64/boot/dts/qcom/sdm845-ayn-odin-m0.dts b/arch/arm64/boot/dts/qcom/sdm845-ayn-odin-m0.dts
new file mode 100644
index 000000000000..54daea10f2e7
--- /dev/null
+++ b/arch/arm64/boot/dts/qcom/sdm845-ayn-odin-m0.dts
@@ -0,0 +1,57 @@
+// SPDX-License-Identifier: GPL-2.0
+/*
+ * Copyright (c) 2024, Anton Bambura.
+ */
+
+/dts-v1/;
+
+#include "sdm845-ayn-odin-common.dtsi"
+
+/ {
+	model = "AYN Odin M0";
+	compatible = "ayn,odin-m0", "ayn,odin", "qcom,sdm845";
+};
+
+&display_panel {
+	status = "okay";
+	compatible = "ayn,odin-m0-panel";
+	rotation = <90>;
+};
+
+&i2c5 {
+	touchscreen@70 {
+		compatible = "syna,rmi4-i2c";
+		reg = <0x70>;
+
+		#address-cells = <1>;
+		#size-cells = <0>;
+
+		reset-gpios = <&tlmm 104 GPIO_ACTIVE_LOW>;
+		interrupts-extended = <&tlmm 125 IRQ_TYPE_EDGE_FALLING>;
+
+		vdd-supply = <&vreg_l28a_3p0>;
+		vio-supply = <&vreg_l14a_1p8>;
+
+		syna,reset-delay-ms = <200>;
+		syna,startup-delay-ms = <200>;
+
+		panel = <&display_panel>;
+
+		pinctrl-names = "default";
+		pinctrl-0 = <&ts_int_active &ts_reset_active>;
+
+		rmi4-f01@1 {
+			reg = <0x01>;
+			syna,nosleep-mode = <1>;
+		};
+
+		rmi4-f12@12 {
+			reg = <0x12>;
+			touchscreen-inverted-x;
+			touchscreen-x-mm = <75>;
+			touchscreen-y-mm = <132>;
+			syna,sensor-type = <1>;
+			syna,rezero-wait-ms = <200>;
+		};
+	};
+};
diff --git a/arch/arm64/boot/dts/qcom/sdm845-ayn-odin-m2.dts b/arch/arm64/boot/dts/qcom/sdm845-ayn-odin-m2.dts
new file mode 100644
index 000000000000..63f8f94916f1
--- /dev/null
+++ b/arch/arm64/boot/dts/qcom/sdm845-ayn-odin-m2.dts
@@ -0,0 +1,57 @@
+// SPDX-License-Identifier: GPL-2.0
+/*
+ * Copyright (c) 2024, Anton Bambura.
+ */
+
+/dts-v1/;
+
+#include "sdm845-ayn-odin-common.dtsi"
+
+/ {
+	model = "AYN Odin M2";
+	compatible = "ayn,odin-m2", "ayn,odin", "qcom,sdm845";
+};
+
+&display_panel {
+	status = "okay";
+	compatible = "ayn,odin-m2-panel";
+	rotation = <270>;
+};
+
+&i2c5 {
+	touchscreen@20 {
+		compatible = "syna,rmi4-i2c";
+		reg = <0x20>;
+
+		#address-cells = <1>;
+		#size-cells = <0>;
+
+		reset-gpios = <&tlmm 104 GPIO_ACTIVE_LOW>;
+		interrupts-extended = <&tlmm 125 IRQ_TYPE_EDGE_FALLING>;
+
+		vdd-supply = <&vreg_l28a_3p0>;
+		vio-supply = <&vreg_l14a_1p8>;
+
+		syna,reset-delay-ms = <200>;
+		syna,startup-delay-ms = <200>;
+
+		panel = <&display_panel>;
+
+		pinctrl-names = "default";
+		pinctrl-0 = <&ts_int_active &ts_reset_active>;
+
+		rmi4-f01@1 {
+			reg = <0x01>;
+			syna,nosleep-mode = <1>;
+		};
+
+		rmi4-f12@12 {
+			reg = <0x12>;
+			touchscreen-inverted-x;
+			touchscreen-x-mm = <75>;
+			touchscreen-y-mm = <132>;
+			syna,sensor-type = <1>;
+			syna,rezero-wait-ms = <200>;
+		};
+	};
+};
diff --git a/arch/arm64/boot/dts/qcom/sdm845.dtsi b/arch/arm64/boot/dts/qcom/sdm845.dtsi
index 4ccd95038d41..2aa5c2bc602a 100644
--- a/arch/arm64/boot/dts/qcom/sdm845.dtsi
+++ b/arch/arm64/boot/dts/qcom/sdm845.dtsi
@@ -5637,7 +5637,7 @@ sound: sound {
 	};
 
 	thermal-zones {
-		cpu0-thermal {
+		cpu0_thermal: cpu0-thermal {
 			polling-delay-passive = <250>;
 
 			thermal-sensors = <&tsens0 1>;
@@ -5663,7 +5663,7 @@ cpu0_crit: cpu-crit {
 			};
 		};
 
-		cpu1-thermal {
+		cpu1_thermal: cpu1-thermal {
 			polling-delay-passive = <250>;
 
 			thermal-sensors = <&tsens0 2>;
@@ -5689,7 +5689,7 @@ cpu1_crit: cpu-crit {
 			};
 		};
 
-		cpu2-thermal {
+		cpu2_thermal: cpu2-thermal {
 			polling-delay-passive = <250>;
 
 			thermal-sensors = <&tsens0 3>;
@@ -5715,7 +5715,7 @@ cpu2_crit: cpu-crit {
 			};
 		};
 
-		cpu3-thermal {
+		cpu3_thermal: cpu3-thermal {
 			polling-delay-passive = <250>;
 
 			thermal-sensors = <&tsens0 4>;
@@ -5741,7 +5741,7 @@ cpu3_crit: cpu-crit {
 			};
 		};
 
-		cpu4-thermal {
+		cpu4_thermal: cpu4-thermal {
 			polling-delay-passive = <250>;
 
 			thermal-sensors = <&tsens0 7>;
@@ -5767,7 +5767,7 @@ cpu4_crit: cpu-crit {
 			};
 		};
 
-		cpu5-thermal {
+		cpu5_thermal: cpu5-thermal {
 			polling-delay-passive = <250>;
 
 			thermal-sensors = <&tsens0 8>;
@@ -5793,7 +5793,7 @@ cpu5_crit: cpu-crit {
 			};
 		};
 
-		cpu6-thermal {
+		cpu6_thermal: cpu6-thermal {
 			polling-delay-passive = <250>;
 
 			thermal-sensors = <&tsens0 9>;
@@ -5819,7 +5819,7 @@ cpu6_crit: cpu-crit {
 			};
 		};
 
-		cpu7-thermal {
+		cpu7_thermal: cpu7-thermal {
 			polling-delay-passive = <250>;
 
 			thermal-sensors = <&tsens0 10>;
@@ -5845,7 +5845,7 @@ cpu7_crit: cpu-crit {
 			};
 		};
 
-		aoss0-thermal {
+		aoss0_thermal: aoss0-thermal {
 			polling-delay-passive = <250>;
 
 			thermal-sensors = <&tsens0 0>;
@@ -5859,7 +5859,7 @@ aoss0_alert0: trip-point0 {
 			};
 		};
 
-		cluster0-thermal {
+		cluster0_thermal: cluster0-thermal {
 			polling-delay-passive = <250>;
 
 			thermal-sensors = <&tsens0 5>;
@@ -5878,7 +5878,7 @@ cluster0_crit: cluster0-crit {
 			};
 		};
 
-		cluster1-thermal {
+		cluster1_thermal: cluster1-thermal {
 			polling-delay-passive = <250>;
 
 			thermal-sensors = <&tsens0 6>;
@@ -5897,7 +5897,7 @@ cluster1_crit: cluster1-crit {
 			};
 		};
 
-		gpu-top-thermal {
+		gpu_top_thermal: gpu-top-thermal {
 			polling-delay-passive = <250>;
 
 			thermal-sensors = <&tsens0 11>;
@@ -5930,7 +5930,7 @@ trip-point2 {
 			};
 		};
 
-		gpu-bottom-thermal {
+		gpu_bottom_thermal: gpu-bottom-thermal {
 			polling-delay-passive = <250>;
 
 			thermal-sensors = <&tsens0 12>;
-- 
GitLab

